In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/scoutium-datas/scoutium_potential_labels.csv
/kaggle/input/scoutium-datas/scoutium_attributes.csv


In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
from sklearn.exceptions import ConvergenceWarning
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier, VotingClassifier
from sklearn.model_selection import GridSearchCV, cross_validate, RandomizedSearchCV, validation_curve
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from catboost import CatBoostClassifier
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.preprocessing import MinMaxScaler, LabelEncoder, StandardScaler, RobustScaler

warnings.simplefilter(action='ignore', category=FutureWarning)
warnings.simplefilter("ignore", category=ConvergenceWarning)

pd.set_option('display.max_columns', None)
# pd.set_option('display.max_rows', None)
pd.set_option('display.width', None)
pd.set_option('display.float_format', lambda x: '%.3f' % x)
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier, VotingClassifier, AdaBoostClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_validate, GridSearchCV
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import StandardScaler
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from catboost import CatBoostClassifier


# Adım 1: scoutium_attributes.csv ve scoutium_potential_labels.csv dosyalarını okutunuz.

In [3]:
df1 = pd.read_csv("/kaggle/input/scoutium-datas/scoutium_attributes.csv",delimiter=";")
df1.head()

task_response_id  match_id  evaluator_id  player_id  position_id  \
0              4915     62935        177676    1361061            2   
1              4915     62935        177676    1361061            2   
2              4915     62935        177676    1361061            2   
3              4915     62935        177676    1361061            2   
4              4915     62935        177676    1361061            2   

   analysis_id  attribute_id  attribute_value  
0     12818495          4322           56.000  
1     12818495          4323           56.000  
2     12818495          4324           67.000  
3     12818495          4325           56.000  
4     12818495          4326           45.000

In [4]:
df2 = pd.read_csv("/kaggle/input/scoutium-datas/scoutium_potential_labels.csv",delimiter=";")

# Adım 2: Okutmuş olduğumuz csv dosyalarını merge fonksiyonunu kullanarak birleştiriniz.

In [5]:
df = pd.merge(df1, df2, on=['task_response_id', 'match_id', 'evaluator_id', 'player_id'])


In [6]:
df.head()

task_response_id  match_id  evaluator_id  player_id  position_id  \
0              4915     62935        177676    1361061            2   
1              4915     62935        177676    1361061            2   
2              4915     62935        177676    1361061            2   
3              4915     62935        177676    1361061            2   
4              4915     62935        177676    1361061            2   

   analysis_id  attribute_id  attribute_value potential_label  
0     12818495          4322           56.000         average  
1     12818495          4323           56.000         average  
2     12818495          4324           67.000         average  
3     12818495          4325           56.000         average  
4     12818495          4326           45.000         average

# Adım 3: position_id içerisindeki Kaleci (1) sınıfını veri setinden kaldırınız.

In [7]:
df = df[df.position_id != 1]
df["position_id"].value_counts()

position_id
2     1972
6     1428
10    1088
8     1020
7      986
3      986
4      884
9      850
5      816
Name: count, dtype: int64

# Adım 4: potential_label içerisindeki below_average sınıfını veri setinden kaldırınız.( below_average sınıfı tüm verisetinin %1'ini oluşturur)

In [8]:
df = df[~df["potential_label"].isin(["below_average"])]
df["potential_label"].unique()

array(['average', 'highlighted'], dtype=object)

# Adım 5: Oluşturduğunuz veri setinden “pivot_table” fonksiyonunu 
# kullanarak bir tablo oluşturunuz. Bu pivot table'da her satırda bir oyuncu
# olacak şekilde manipülasyon yapınız.
# Adım 5.1: İndekste “player_id”,“position_id” ve “potential_label”, 
# sütunlarda “attribute_id” ve değerlerde scout’ların oyunculara verdiği puan
# “attribute_value” olacak şekilde pivot table’ı oluşturunuz.
# Adım 2: “reset_index” fonksiyonunu kullanarak indeksleri değişken olarak 
# atayınız ve “attribute_id” sütunlarının isimlerini stringe çeviriniz.

In [9]:
df_table = df.pivot_table("attribute_value", ["player_id", "position_id" , "potential_label"] , "attribute_id")

df_table.head()

attribute_id                            4322   4323   4324   4325   4326  \
player_id position_id potential_label                                      
1355710   7           average         50.500 50.500 34.000 50.500 45.000   
1356362   9           average         67.000 67.000 67.000 67.000 67.000   
1356375   3           average         67.000 67.000 67.000 67.000 67.000   
          4           average         67.000 78.000 67.000 67.000 67.000   
1356411   9           average         67.000 67.000 78.000 78.000 67.000   

attribute_id                            4327   4328   4329   4330   4332  \
player_id position_id potential_label                                      
1355710   7           average         45.000 45.000 45.000 50.500 56.000   
1356362   9           average         67.000 67.000 67.000 67.000 67.000   
1356375   3           average         67.000 67.000 78.000 67.000 67.000   
          4           average         78.000 78.000 78.000 56.000 67.000   
1356411   9           average         67.000 67.000 67.000 89.000 78.000   

attribute_id                            4333   4335   4338   4339   4340  \
player_id position_id potential_label                                      
1355710   7           average         39.500 34.000 39.500 39.500 45.000   
1356362   9           average         67.000 67.000 67.000 67.000 67.000   
1356375   3           average         78.000 56.000 67.000 67.000 67.000   
          4           average         67.000 67.000 78.000 78.000 56.000   
1356411   9           average         67.000 67.000 67.000 56.000 56.000   

attribute_id                            4341   4342   4343   4344   4345  \
player_id position_id potential_label                                      
1355710   7           average         45.000 50.500 28.500 23.000 39.500   
1356362   9           average         67.000 67.000 67.000 56.000 67.000   
1356375   3           average         67.000 67.000 56.000 56.000 67.000   
          4           average         67.000 67.000 45.000 45.000 56.000   
1356411   9           average         67.000 78.000 56.000 56.000 67.000   

attribute_id                            4348   4349   4350   4351   4352  \
player_id position_id potential_label                                      
1355710   7           average         28.500 28.500 45.000 50.500 56.000   
1356362   9           average         67.000 56.000 67.000 67.000 67.000   
1356375   3           average         67.000 56.000 56.000 67.000 67.000   
          4           average         67.000 67.000 67.000 67.000 78.000   
1356411   9           average         56.000 67.000 56.000 67.000 67.000   

attribute_id                            4353   4354   4355   4356   4357  \
player_id position_id potential_label                                      
1355710   7           average         34.000 39.500 50.500 34.000 34.000   
1356362   9           average         67.000 78.000 67.000 67.000 67.000   
1356375   3           average         67.000 78.000 67.000 67.000 67.000   
          4           average         67.000 67.000 67.000 56.000 67.000   
1356411   9           average         56.000 67.000 67.000 56.000 67.000   

attribute_id                            4407   4408   4423   4426  
player_id position_id potential_label                              
1355710   7           average         56.000 34.000 34.000 56.000  
1356362   9           average         67.000 67.000 56.000 78.000  
1356375   3           average         67.000 67.000 56.000 78.000  
          4           average         56.000 67.000 45.000 56.000  
1356411   9           average         89.000 56.000 67.000 78.000

In [10]:
df_table = df_table.reset_index()

In [11]:
df_table.head()

attribute_id  player_id  position_id potential_label   4322   4323   4324  \
0               1355710            7         average 50.500 50.500 34.000   
1               1356362            9         average 67.000 67.000 67.000   
2               1356375            3         average 67.000 67.000 67.000   
3               1356375            4         average 67.000 78.000 67.000   
4               1356411            9         average 67.000 67.000 78.000   

attribute_id   4325   4326   4327   4328   4329   4330   4332   4333   4335  \
0            50.500 45.000 45.000 45.000 45.000 50.500 56.000 39.500 34.000   
1            67.000 67.000 67.000 67.000 67.000 67.000 67.000 67.000 67.000   
2            67.000 67.000 67.000 67.000 78.000 67.000 67.000 78.000 56.000   
3            67.000 67.000 78.000 78.000 78.000 56.000 67.000 67.000 67.000   
4            78.000 67.000 67.000 67.000 67.000 89.000 78.000 67.000 67.000   

attribute_id   4338   4339   4340   4341   4342   4343   4344   4345   4348  \
0            39.500 39.500 45.000 45.000 50.500 28.500 23.000 39.500 28.500   
1            67.000 67.000 67.000 67.000 67.000 67.000 56.000 67.000 67.000   
2            67.000 67.000 67.000 67.000 67.000 56.000 56.000 67.000 67.000   
3            78.000 78.000 56.000 67.000 67.000 45.000 45.000 56.000 67.000   
4            67.000 56.000 56.000 67.000 78.000 56.000 56.000 67.000 56.000   

attribute_id   4349   4350   4351   4352   4353   4354   4355   4356   4357  \
0            28.500 45.000 50.500 56.000 34.000 39.500 50.500 34.000 34.000   
1            56.000 67.000 67.000 67.000 67.000 78.000 67.000 67.000 67.000   
2            56.000 56.000 67.000 67.000 67.000 78.000 67.000 67.000 67.000   
3            67.000 67.000 67.000 78.000 67.000 67.000 67.000 56.000 67.000   
4            67.000 56.000 67.000 67.000 56.000 67.000 67.000 56.000 67.000   

attribute_id   4407   4408   4423   4426  
0            56.000 34.000 34.000 56.000  
1            67.000 67.000 56.000 78.000  
2            67.000 67.000 56.000 78.000  
3            56.000 67.000 45.000 56.000  
4            89.000 56.000 67.000 78.000

In [12]:
df_table.columns = df_table.columns.map(str)

# Adım 6: Label Encoder fonksiyonunu kullanarak “potential_label” 
# kategorilerini (average, highlighted) sayısal olarak ifade ediniz.

In [13]:
def label_encoder(dataframe, binary_col):
    labelencoder = LabelEncoder()
    dataframe[binary_col] = labelencoder.fit_transform(dataframe[binary_col])
    return dataframe

binary_cols = [col for col in df_table.columns if col in "potential_label"]

for col in binary_cols:
    label_encoder(df_table, col)

df_table.head()

attribute_id  player_id  position_id  potential_label   4322   4323   4324  \
0               1355710            7                0 50.500 50.500 34.000   
1               1356362            9                0 67.000 67.000 67.000   
2               1356375            3                0 67.000 67.000 67.000   
3               1356375            4                0 67.000 78.000 67.000   
4               1356411            9                0 67.000 67.000 78.000   

attribute_id   4325   4326   4327   4328   4329   4330   4332   4333   4335  \
0            50.500 45.000 45.000 45.000 45.000 50.500 56.000 39.500 34.000   
1            67.000 67.000 67.000 67.000 67.000 67.000 67.000 67.000 67.000   
2            67.000 67.000 67.000 67.000 78.000 67.000 67.000 78.000 56.000   
3            67.000 67.000 78.000 78.000 78.000 56.000 67.000 67.000 67.000   
4            78.000 67.000 67.000 67.000 67.000 89.000 78.000 67.000 67.000   

attribute_id   4338   4339   4340   4341   4342   4343   4344   4345   4348  \
0            39.500 39.500 45.000 45.000 50.500 28.500 23.000 39.500 28.500   
1            67.000 67.000 67.000 67.000 67.000 67.000 56.000 67.000 67.000   
2            67.000 67.000 67.000 67.000 67.000 56.000 56.000 67.000 67.000   
3            78.000 78.000 56.000 67.000 67.000 45.000 45.000 56.000 67.000   
4            67.000 56.000 56.000 67.000 78.000 56.000 56.000 67.000 56.000   

attribute_id   4349   4350   4351   4352   4353   4354   4355   4356   4357  \
0            28.500 45.000 50.500 56.000 34.000 39.500 50.500 34.000 34.000   
1            56.000 67.000 67.000 67.000 67.000 78.000 67.000 67.000 67.000   
2            56.000 56.000 67.000 67.000 67.000 78.000 67.000 67.000 67.000   
3            67.000 67.000 67.000 78.000 67.000 67.000 67.000 56.000 67.000   
4            67.000 56.000 67.000 67.000 56.000 67.000 67.000 56.000 67.000   

attribute_id   4407   4408   4423   4426  
0            56.000 34.000 34.000 56.000  
1            67.000 67.000 56.000 78.000  
2            67.000 67.000 56.000 78.000  
3            56.000 67.000 45.000 56.000  
4            89.000 56.000 67.000 78.000

In [14]:
df_table["potential_label"].value_counts()

potential_label
0    215
1     56
Name: count, dtype: int64

# Adım 7: Sayısal değişken kolonlarını “num_cols” adıyla bir listeye atayınız.

In [15]:
def grab_col_names(dataframe, cat_th=10, car_th=20):
    """
    grab_col_names for given dataframe

    :param dataframe:
    :param cat_th:
    :param car_th:
    :return:
    """

    cat_cols = [col for col in dataframe.columns if dataframe[col].dtypes == "O"]

    num_but_cat = [col for col in dataframe.columns if dataframe[col].nunique() < cat_th and
                   dataframe[col].dtypes != "O"]

    cat_but_car = [col for col in dataframe.columns if dataframe[col].nunique() > car_th and
                   dataframe[col].dtypes == "O"]

    cat_cols = cat_cols + num_but_cat
    cat_cols = [col for col in cat_cols if col not in cat_but_car]

    num_cols = [col for col in dataframe.columns if dataframe[col].dtypes != "O"]
    num_cols = [col for col in num_cols if col not in num_but_cat]

    print(f"Observations: {dataframe.shape[0]}")
    print(f"Variables: {dataframe.shape[1]}")
    print(f'cat_cols: {len(cat_cols)}')
    print(f'num_cols: {len(num_cols)}')
    print(f'cat_but_car: {len(cat_but_car)}')
    print(f'num_but_cat: {len(num_but_cat)}')

    # cat_cols + num_cols + cat_but_car = değişken sayısı.
    # num_but_cat cat_cols'un içerisinde zaten.
    # dolayısıyla tüm şu 3 liste ile tüm değişkenler seçilmiş olacaktır: cat_cols + num_cols + cat_but_car
    # num_but_cat sadece raporlama için verilmiştir.

    return cat_cols, cat_but_car, num_cols

In [16]:
cat_cols, cat_but_car, num_cols = grab_col_names(df_table)

Observations: 271
Variables: 37
cat_cols: 7
num_cols: 30
cat_but_car: 0
num_but_cat: 7


In [17]:
num_cols

['player_id',
 '4322',
 '4323',
 '4325',
 '4326',
 '4327',
 '4329',
 '4330',
 '4332',
 '4333',
 '4335',
 '4338',
 '4339',
 '4340',
 '4341',
 '4342',
 '4343',
 '4344',
 '4345',
 '4348',
 '4349',
 '4350',
 '4351',
 '4353',
 '4354',
 '4355',
 '4356',
 '4407',
 '4408',
 '4426']

Player id nümerik değil o yüzden onu çıkardık

In [18]:
num_cols = [col for col in num_cols if col not in "player_id"]

# Adım 8: Kaydettiğiniz bütün “num_cols” değişkenlerindeki veriyi ölçeklendirmek için StandardScaler uygulayınız.

In [19]:
scaler = StandardScaler()
df_table[num_cols] = scaler.fit_transform(df_table[num_cols])
df_table[num_cols].head()

attribute_id   4322   4323   4325   4326   4327   4329   4330   4332   4333  \
0            -0.543 -0.559 -0.438 -0.767 -0.795 -0.792 -0.446 -0.123 -1.224   
1             0.595  0.561  0.683  0.723  0.723  0.404  0.728  0.691  0.459   
2             0.595  0.561  0.683  0.723  0.723  1.002  0.728  0.691  1.132   
3             0.595  1.308  0.683  0.723  1.482  1.002 -0.055  0.691  0.459   
4             0.595  0.561  1.430  0.723  0.723  0.404  2.294  1.505  0.459   

attribute_id   4335   4338   4339   4340   4341   4342   4343   4344   4345  \
0            -1.036 -1.126 -1.009 -0.542 -0.690 -0.535 -1.067 -1.206 -1.005   
1             0.996  0.632  0.840  0.912  0.799  0.588  1.230  0.750  0.742   
2             0.319  0.632  0.840  0.912  0.799  0.588  0.574  0.750  0.742   
3             0.996  1.335  1.580  0.185  0.799  0.588 -0.083  0.098  0.043   
4             0.996  0.632  0.101  0.185  0.799  1.337  0.574  0.750  0.742   

attribute_id   4348   4349   4350   4351   4353   4354   4355   4356   4407  \
0            -1.314 -1.042 -0.693 -0.436 -1.282 -1.132 -0.495 -1.235 -0.143   
1             0.879  0.670  0.848  0.717  0.814  1.053  0.632  0.915  0.530   
2             0.879  0.670  0.077  0.717  0.814  1.053  0.632  0.915  0.530   
3             0.879  1.355  0.848  0.717  0.814  0.428  0.632  0.198 -0.143   
4             0.252  1.355  0.077  0.717  0.115  0.428  0.632  0.198  1.874   

attribute_id   4408   4426  
0            -1.487 -0.253  
1             0.669  1.042  
2             0.669  1.042  
3             0.669 -0.253  
4            -0.050  1.042

# Adım 9: Elimizdeki veri seti üzerinden minimum hata ile futbolcuların potansiyel etiketlerini tahmin eden bir makine öğrenmesi modeli geliştiriniz. (Roc_auc, f1, precision, recall, accuracy metriklerini yazdırınız.)


In [20]:
y = df_table["potential_label"]
X = df_table.drop(["potential_label", "position_id", "player_id"], axis=1)

#Base model

In [21]:
def base_models(X, y, scoring="roc_auc"):
    print("Base Models....")
    classifiers = [('LR', LogisticRegression()),
                   ('KNN', KNeighborsClassifier()),
                   ("SVC", SVC()),
                   ("CART", DecisionTreeClassifier()),
                   ("RF", RandomForestClassifier()),
                   ('Adaboost', AdaBoostClassifier()),
                   ('GBM', GradientBoostingClassifier()),
                   ('XGBoost', XGBClassifier(use_label_encoder=False, eval_metric='logloss')),
                   ('LightGBM', LGBMClassifier()),
                   # ('CatBoost', CatBoostClassifier(verbose=False))
                   ]

    for name, classifier in classifiers:
        cv_results = cross_validate(classifier, X, y, cv=3, scoring=scoring)
        print(f"{scoring}: {round(cv_results['test_score'].mean(), 4)} ({name}) ")

base_models(X, y, scoring="accuracy")

Base Models....
accuracy: 0.8486 (LR) 
accuracy: 0.8044 (KNN) 
accuracy: 0.7934 (SVC) 
accuracy: 0.8193 (CART) 
accuracy: 0.8781 (RF) 
accuracy: 0.823 (Adaboost) 
accuracy: 0.8414 (GBM) 
accuracy: 0.8562 (XGBoost) 
accuracy: 0.8598 (LightGBM) 


Base Models....
accuracy: 0.8486 (LR) 
accuracy: 0.8044 (KNN) 
accuracy: 0.7934 (SVC) 
accuracy: 0.8046 (CART) 
accuracy: 0.867 (RF) 
accuracy: 0.823 (Adaboost) 
accuracy: 0.8488 (GBM) 
accuracy: 0.8562 (XGBoost) 
accuracy: 0.8598 (LightGBM) 



BEN 
accuracy: 0.867 (RF) 
accuracy: 0.8562 (XGBoost) 
accuracy: 0.8598 (LightGBM) 
seçtim

RF 

In [22]:
rf_model = RandomForestClassifier(random_state=17)

rf_params = {"max_depth": [8, None],
             "max_features": [5, 7, "auto"],
             "min_samples_split": [15, 20],
             "n_estimators": [200, 300]}

rf_best_grid = GridSearchCV(rf_model, rf_params, cv=3, n_jobs=-1, verbose=True).fit(X, y)

rf_best_grid.best_params_

rf_final = rf_model.set_params(**rf_best_grid.best_params_, random_state=17).fit(X, y)

cv_results = cross_validate(rf_final, X, y, cv=3, scoring=["accuracy", "f1", "roc_auc"])
cv_results['test_accuracy'].mean()

Fitting 3 folds for each of 24 candidates, totalling 72 fits


0.8817663817663818

In [23]:
cv_results['test_f1'].mean()

0.5934682428371068

In [24]:
cv_results['test_roc_auc'].mean()

0.887351970824296

XGBOOST

In [25]:
xgboost_model = XGBClassifier(random_state=17)

xgboost_params = {"learning_rate": [0.1, 0.01],
                  "max_depth": [5, 8],
                  "n_estimators": [100, 200]}
xgboost_best_grid = GridSearchCV(xgboost_model, xgboost_params, cv=5, n_jobs=-1, verbose=True).fit(X, y)

xgboost_final = xgboost_model.set_params(**xgboost_best_grid.best_params_, random_state=17).fit(X, y)

cv_results = cross_validate(xgboost_final, X, y, cv=5, scoring=["accuracy", "f1", "roc_auc"])
cv_results['test_accuracy'].mean()
cv_results['test_f1'].mean()
cv_results['test_roc_auc'].mean()

Fitting 5 folds for each of 8 candidates, totalling 40 fits


0.873784355179704

LIGHTGBM

In [26]:
lgbm_model = LGBMClassifier(random_state=17)

lightgbm_params = {"learning_rate": [0.01, 0.1],
                   "n_estimators": [300, 500]}

lgbm_best_grid = GridSearchCV(lgbm_model, lgbm_params, cv=5, n_jobs=-1, verbose=True).fit(X, y)

lgbm_final = lgbm_model.set_params(**lgbm_best_grid.best_params_, random_state=17).fit(X, y)

cv_results = cross_validate(lgbm_final, X, y, cv=5, scoring=["accuracy", "f1", "roc_auc"])

cv_results['test_accuracy'].mean()
cv_results['test_f1'].mean()
cv_results['test_roc_auc'].mean()

NameError: name 'lgbm_params' is not defined

# Adım 10: Değişkenlerin önem düzeyini belirten feature_importance fonksiyonunu kullanarak özelliklerin sıralamasını çizdiriniz.

In [ ]:
def plot_importance(model, features, num=len(X), save=False):
    feature_imp = pd.DataFrame({'Value': model.feature_importances_, 'Feature': features.columns})
    plt.figure(figsize=(10, 10))
    sns.set(font_scale=1)
    sns.barplot(x="Value", y="Feature", data=feature_imp.sort_values(by="Value",
                                                                     ascending=False)[0:num])
    plt.title('Features')
    plt.tight_layout()
    plt.show(block=True)
    if save:
        plt.savefig('importances.png')

plot_importance(rf_final, X)
plot_importance(lgbm_final, X)
plot_importance(xgboost_final, X)